src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public String getFormat() {
+        return isPublicKey ? "X.509" :  "PKCS#8";
+    }
+
+    @Override
+    public byte[] getEncoded() {
+        initEncoded();
+        return encoded != null ? encoded.clone() : encoded;
+    }
+
+    protected synchronized void initEncoded() {
+        if (encoded == null) {
+            encoded = isPublicKey ? use(EvpKey::encodePublicKey) : use(EvpKey::encodePrivateKey);
+        }
+    }
+
+    protected <X extends Throwable> BigInteger nativeBN(final MiscInterfaces.ThrowingLongFunction<byte[], X> fn)
+            throws X {
+        byte[] raw = use(fn::apply);
+        return new BigInteger(1, raw);
+    }
+
+    protected <T extends AlgorithmParameterSpec> T nativeParams(final Class<T> paramSpec) {
+        byte[] encodedParams = use(EvpKey::getDerEncodedParams);
+        try {
+            AlgorithmParameters params = AlgorithmParameters.getInstance(type.jceName);
+            params.init(encodedParams);
+            return params.getParameterSpec(paramSpec);
+        } catch (final GeneralSecurityException | IOException ex) {
+            throw new RuntimeCryptoException("Unable to deserialize parameters: "
+                + Base64.getEncoder().encodeToString(encoded), ex);
+        }
+    }
+
+    /**
+     * This method will be called by @{link #destroy()} after calling @{code internalKey.release()}.
+     */
+    protected synchronized void destroyJavaState() {
+        // NOP
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        // We try to avoid comparing the encoded values
+        // because it may be slow and may pull secret data into the Java heap
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof Key)) { // Implicit null check
+            return false;
+        }
+        final Key other = (Key) obj;
+        if (!getAlgorithm().equalsIgnoreCase(other.getAlgorithm())) {
+            return false;
+        }
+
+        final byte[] otherEncoded;
+        if (obj.getClass().equals(getClass())) {
+            // If it is also an EvpKey then we can see if the internal key is the same
+            EvpKey evpOther = (EvpKey) obj;
+            if (internalKey.equals(evpOther.internalKey)) {
+                return true;
+            }
+            // If it is also an EvpKey then we can grab the other encoded value without copying it
+            evpOther.initEncoded();
+            otherEncoded = evpOther.encoded;
+        } else {
+            otherEncoded = other.getEncoded();
+        }
+
+        // Constant time equality check
+        initEncoded();
+        return MessageDigest.isEqual(encoded, otherEncoded);
+    }
+
+    @Override
+    public int hashCode() {
+        // TODO: Consider ways to avoid exposing the entire encoded object ot Java for private keys just for a hashCode
+        if (cachedHashCode == null) {
+            synchronized (this) {
+                initEncoded();
```

Comment:
I do this to avoid having multiple threads calculating the cached hash code simultaneously.

Reply:
Since the synchronized block is inside the `null` check, there could still potentially be multiple threads reaching the start of this critical section, and each thread would still compute its own hash and cache it.

Reply:
This is fixed in the newest version.